src/EditorFeatures/Core/Implementation/AddImports/AddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,120 @@
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
+
+            if (!args.SubjectBuffer.GetFeatureOnOffOption(FeatureOnOffOptions.AddImportsOnPaste))
+            {
+                return;
+            }
+
+            // Create a tracking span from the pre-paste caret position that will grow as text is inserted.
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            // Applying the post-paste snapshot to the tracking span gives us the span of pasted text.
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = TextSpan.FromBounds(snapshotSpan.Start, snapshotSpan.End);
+
+            AddMissingImportsForPaste(args, executionContext, textSpan);
+        }
+
+        public CommandState GetCommandState(PasteCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        private void AddMissingImportsForPaste(PasteCommandArgs args, CommandExecutionContext executionContext, TextSpan textSpan)
+        {
+            var sourceTextContainer = args.SubjectBuffer.AsTextContainer();
+            if (!Workspace.TryGetWorkspace(sourceTextContainer, out var workspace))
+            {
+                return;
+            }
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(sourceTextContainer);
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+
+            if (document is null)
+            {
+                return;
+            }
+
+            using var _ = executionContext.OperationContext.AddScope(allowCancellation: true, EditorFeaturesResources.Adding_missing_import_directives);
+            var cancellationToken = executionContext.OperationContext.UserCancellationToken;
+
+            var updatedProject = _threadingContext.JoinableTaskFactory.Run(() => AddMissingImportsAsync(document, textSpan, cancellationToken));
```

Comment:
so... it's unclear to me why we have this as blocking at all.  Why not just kick off the work, and then just track if the buffer gets modified otherwise. If so, cancel the work.

This way, we can ensure that if users paste and just keep typing, we don't interfere with them.  But if they're willing to wait, they can get the imports.

Reply:
This follows VS design guidelines for when to show a wait dialog ([internal doc](https://devdivdesignguide.azurewebsites.net/windows/notifications-and-progress/progress-indicators.html#factors)) and went through office hours with the design team to confirm this was the right thing to do. 

Some highlights about not showing a dialog: 

1. How does the user know something is happening?
2. How long should I wait for usings/imports to be added? 
3. How do I know it failed vs me accidently doing something I shouldn't have?